cpp/bench/ann/src/common/ann_types.hpp

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "cuda_stub.hpp"  // cudaStream_t
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace cuvs::bench::ann {
+
+enum Objective {
+  THROUGHPUT,  // See how many vectors we can push through
+  LATENCY      // See how fast we can push a vector through
+};
+
+enum class MemoryType {
+  Host,
+  HostMmap,
+  Device,
+};
+
+enum class Metric {
+  kInnerProduct,
+  kEuclidean,
+};
+
+inline auto parse_metric(const std::string& metric_str) -> Metric
+{
+  if (metric_str == "inner_product") {
+    return cuvs::bench::ann::Metric::kInnerProduct;
+  } else if (metric_str == "euclidean") {
+    return cuvs::bench::ann::Metric::kEuclidean;
+  } else {
+    throw std::runtime_error("invalid metric: '" + metric_str + "'");
+  }
+}
+
+inline auto parse_memory_type(const std::string& memory_type) -> MemoryType
+{
+  if (memory_type == "host") {
+    return MemoryType::Host;
+  } else if (memory_type == "mmap") {
+    return MemoryType::HostMmap;
+  } else if (memory_type == "device") {
+    return MemoryType::Device;
+  } else {
+    throw std::runtime_error("invalid memory type: '" + memory_type + "'");
+  }
+}
+
+struct AlgoProperty {
+  MemoryType dataset_memory_type;
+  // neighbors/distances should have same memory type as queries
+  MemoryType query_memory_type;
+};
+
+class AnnBase {
```

Comment:
Can we maybe refer to this as an "Index"?

Reply:
For now I renamed this to `algo_base` and `algo<T>`. I hesitate changing this to `index`, because this would clash with actual `index` types from other namespaces, which are members of these wrappers. What do you think about this?